devtools/x-core/src/git.rs

```diff
@@ -0,0 +1,207 @@
+    pub fn merge_base(&self, commit_ref: &str) -> Result<GitHash> {
+        let output = self
+            .git_command()
+            .args(&["merge-base", "HEAD", commit_ref])
+            .output()
+            .map_err(|err| {
+                SystemError::io(format!("running git merge-base HEAD {}", commit_ref), err)
+            })?;
+        if !output.status.success() {
+            return Err(SystemError::Exec {
+                cmd: "git merge-base",
+                status: output.status,
+            });
+        }
+
+        // The output is a hex-encoded hash followed by a newline.
+        let stdout = &output.stdout[..(output.stdout.len() - 1)];
+        GitHash::from_hex(stdout)
+    }
+
+    /// Returns the files changed between the given commits, or the current directory if the new
+    /// commit isn't specified.
+    ///
+    /// For more about the diff filter, see `man git-diff`'s help for `--diff-filter`.
+    pub fn files_changed_between<'a>(
+        &self,
+        old: impl Into<Cow<'a, OsStr>>,
+        new: impl Into<Option<Cow<'a, OsStr>>>,
+        // TODO: make this more well-typed/express more of the diff model in Rust
+        diff_filter: Option<&str>,
+    ) -> Result<Paths0> {
+        let mut command = self.git_command();
+        command.args(&["diff", "-z", "--name-only"]);
+        if let Some(diff_filter) = diff_filter {
+            command.arg(format!("--diff-filter={}", diff_filter));
+        }
+        command.arg(old.into());
+        if let Some(new) = new.into() {
+            command.arg(new);
+        }
+
+        let output = command
+            .output()
+            .map_err(|err| SystemError::io("running git diff", err))?;
+        if !output.status.success() {
+            return Err(SystemError::Exec {
+                cmd: "git diff",
+                status: output.status,
+            });
+        }
+
+        Ok(Paths0::new_unix(output.stdout))
+    }
+
+    /// Returns a package graph for the given commit, using a scratch repo if necessary.
+    pub fn package_graph_at(&self, commit_ref: &GitHash) -> Result<PackageGraph> {
+        // Create or initialize the scratch worktree.
+        let scratch = self.get_or_init_scratch(commit_ref)?;
+
+        // Compute the package graph for the scratch worktree.
+        Ok(MetadataCommand::new().current_dir(scratch).build_graph()?)
+    }
+
+    // ---
+    // Helper methods
+    // ---
+
+    // TODO: abstract out command running and error handling
+    fn git_command(&self) -> Command {
+        // TODO: add support for the GIT environment variable?
+        let mut command = Command::new("git");
+        command.current_dir(self.root).stderr(Stdio::inherit());
+        command
+    }
+
+    fn get_or_init_scratch(&self, hash: &GitHash) -> Result<PathBuf> {
+        let mut scratch_dir = self.root.join("target");
+        scratch_dir.extend(&["x-scratch", "tree"]);
```

Comment:
can we get a random string location appended to x-scratch here?

Reply:
I actually thought about that and decided not to do it that way -- a persistent scratch directory maintained by x is what I ended up going with. Means in the common case the same directory can be reused locally.

Reply:
Comment the function to that effect?

Reply:
Done.